srv/server.go

```diff
@@ -0,0 +1,371 @@
+	job_id := in.GetJobId()
+	username, err := getUsername(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	id := uuid.New()
+	if err = id.UnmarshalBinary(job_id); err != nil {
+		return nil, err
+	}
+
+	status, err := s.statusOfJob(username, id)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.StatusResponse{
+		JobId:    job_id,
+		Status:   status.String(),
+		ExitCode: int32(status.ExitCode),
+	}, nil
+}
+
+// Output receives an RPC containing  job id of the job
+// On success, it will start streaming output of the job to the client
+// On failure, it will return an error
+func (s *server) Output(in *pb.OutputRequest, stream pb.Worker_OutputServer) (err error) {
+	job_id := in.GetJobId()
+	username, err := getUsername(stream.Context())
+	if err != nil {
+		return err
+	}
+
+	id := uuid.New()
+	if err = id.UnmarshalBinary(job_id); err != nil {
+		return err
+	}
+
+	streamC, err := s.outputOfJob(username, id)
+	if err != nil {
+		return err
+	}
+
+	// try sending up to 10 lines at a time
+	// if stream channel does not have data within 1 millisecond, send whatever we already have
+	maxLinesPerRPC := 10
+	maxWaitTime := time.Millisecond
+	for {
+		// read until channel is closed, which indicates end of the job
+		batch, end := getNextLineBatch(streamC, maxLinesPerRPC, maxWaitTime)
+		// if batch has length of 0, but stream has not ended, then try again
+		if len(batch) == 0 && !end {
+			continue
+		}
+		resp := &pb.OutputResponse{
+			OutputLine: batch,
+		}
+
+		if err := stream.Send(resp); err != nil {
+			return err
+		}
+
+		if end {
+			break
+		}
+	}
+
+	return nil
+}
+
+// createJob creates a job with command for user with username
+func (s *server) startJob(username string, command string) (*lib.Job, error) {
+	job := lib.NewJob(command)
+
+	if job == nil {
+		return nil, errors.New("failed to create a job")
+	}
+
+	// Special check whether UUID exists already. Disallow if so as we
+	// should never hit this (almost never: see uuid.NewRandom())
+	// TODO: Should handle this possibility by checking the existing list
+	// of jobs. As jobs are not managed by the library, I will skip this
+	// handling for now. Might need to handle the jobs internally in library
+	// to resolve this in a cleaner fashion.
+	_, exists := s.jobs[job.UUID]
+	if exists {
+		return nil, errors.New("server had a hick-up. Can you try again?")
+	}
+
+	// set the username for this job. All other RPCs will check this field
+	// for authorizing the requested service for this job
+	job.SetUsername(username)
+	s.jobs[job.UUID] = job
+	if err := job.Start(); err != nil {
+		return job, err
+	}
+
+	return job, nil
+}
+
+// stopJob attempts to stop the job with job_id. if username does not match
+// the username in this job, it will return an authorization error.
+func (s *server) stopJob(username string, job_id uuid.UUID) error {
+	authErr := s.authorize(username, job_id)
+	if authErr != nil {
+		return authErr
+	}
+
+	job := s.jobs[job_id]
+	err := job.Stop()
```

Comment:
This would be a nil pointer dereference if the job ID doesn't exist, right?

Reply:
authorize() method would return an error if job didn't exist. so, it is guaranteed that job with job_id exist at this line